src/CanvasRenderingContext2d.cc

```diff
@@ -654,6 +678,125 @@ NAN_METHOD(Context2d::PutImageData) {
 }
 
 /*
+ * Get image data.
+ *
+ *  - sx, sy, sw, sh
+ *
+ */
+
+NAN_METHOD(Context2d::GetImageData) {
+  NanScope();
+
+  Context2d *context = ObjectWrap::Unwrap<Context2d>(args.This());
+  Canvas *canvas = context->canvas();
+
+  int sx = args[0]->Int32Value();
+  int sy = args[1]->Int32Value();
+  int sw = args[2]->Int32Value();
+  int sh = args[3]->Int32Value();
+
+  if (!sw)
+    return NanThrowError("IndexSizeError: The source width is 0.");
+  if (!sh)
+    return NanThrowError("IndexSizeError: The source height is 0.");
+
+  // WebKit and Firefox have this behavior:
+  // Flip the coordinates so the origin is top/left-most:
+  if (sw < 0) {
+    sx += sw;
+    sw = -sw;
+  }
+  if (sh < 0) {
+    sy += sh;
+    sh = -sh;
+  }
+
+  if (sx + sw > canvas->width) sw = canvas->width - sx;
+  if (sy + sh > canvas->height) sh = canvas->height - sy;
+
+  // WebKit/moz functionality. node-canvas used to return in either case.
+  if (sw <= 0) sw = 1;
+  if (sh <= 0) sh = 1;
+
+  // Non-compliant. "Pixels outside the canvas must be returned as transparent
+  // black." This instead clips the returned array to the canvas area.
+  if (sx < 0) {
+    sw += sx;
+    sx = 0;
+  }
+  if (sy < 0) {
+    sh += sy;
+    sy = 0;
+  }
+
+  int size = sw * sh * 4;
+
+  int srcStride = canvas->stride();
+  int dstStride = sw * 4;
+
+  uint8_t *src = canvas->data();
+  uint8_t *dst = (uint8_t *)calloc(1, size);
+  NanAdjustExternalMemory(size);
+
+#if NODE_MAJOR_VERSION == 0 && NODE_MINOR_VERSION <= 10
+  Local<Object> global = Context::GetCurrent()->Global();
+
+  Handle<Value> bufargv[] = { NanNew(size) };
+  Local<Object> buffer = global->Get(NanNew("ArrayBuffer")).As<Function>()->NewInstance(1, bufargv);
+
+  Handle<Value> caargv[] = { buffer, NanNew(0), NanNew(size) };
+  Local<Object> clampedArray = global->Get(NanNew("Uint8ClampedArray")).As<Function>()->NewInstance(3, caargv);
+
+  clampedArray->SetIndexedPropertiesToExternalArrayData(dst, kExternalPixelArray, size);
+#else
+  Local<ArrayBuffer> buffer = ArrayBuffer::New(Isolate::GetCurrent(), size);
+  Local<Uint8ClampedArray> clampedArray = Uint8ClampedArray::New(buffer, 0, size);
+#if NODE_MAJOR_VERSION < 3 // includes node 0.12
+  clampedArray->SetIndexedPropertiesToExternalArrayData(dst, kExternalUint8ClampedArray, size);
+#else
+  // TODO
+#endif
```

Comment:
Instead of doing nothing if `NODE_MAJOR_VERSION < 3` I would rather that we gave a compiler error. Basically, I would prefer it if line 754, 756, 757 and 758 where gone.

Makes sense?